### PR TITLE
Infra/treaty memory

### LINKED ISSUE #135: Add memory of treaties to baseAgent
Add a infinitely sized memory for active treaties called `ActiveTreaties`. 
This will be a map of `string` (a uuid) to `Treaty`.
This will go in baseAgent.

We want to have a few helper functions:

- `AcceptTreaty`
- `DeclineTreaty`
- `FilterExpiredTreaties`

### PR DESCRIPTION
**Pull request by woonmoon**
Closes #135.

This PR adds a map of `activeTreaties` to the base agent, and also updates the duration of each treaty at the end of a day, and removes treaties which have remaining duration of 0.

All message-related code in base agent has been moved to `baseAgentMessage.go`.

The `HandleProposeTreaty()` and `HandleTreatyResponse()` functions have now been added to base agent as well: these have been implemented with a sample implementation.

Note that `HandleProposeTreaty()` automatically rejects treaties - agent teams should override this with their desired behaviour. 

This also fixes the interfaces for some messages - the Reply() function has been updated (see #190).

### TESTING
**Testing Config File**
```
{
    "UseFoodPerAgentRatio": false,
    "FoodOnPlatform": 100,
    "FoodPerAgentRatio": 0,
    "Team1AgtOne": 2,
    "Team1AgtTwo": 2,
    "Team2Agents": 0,
    "Team3Agents": 0,
    "Team4Agents": 0,
    "Team5Agents": 0,
    "Team6Agents": 0,
    "Team7Agent1": 0,
    "RandomAgents": 0,
    "AgentHP": 100,
    "AgentsPerFloor": 1,
    "TicksPerFloor": 10,
    "SimDays": 8,
    "ReshuffleDays": 1,
    "maxHP": 100,
    "weakLevel": 10,
    "width": 48,
    "tau": 15,
    "hpReqCToW": 5,
    "hpCritical": 3,
    "maxDayCritical": 3,
    "HPLossBase": 5,
    "HPLossSlope": 0.2,
    "LogMain": true,
    "SimTimeoutSeconds": 60
}
```
With `SimEnv` in `simulation.go`
```
SimEnv{
	FoodOnPlatform: parameters.FoodOnPlatform,
	AgentCount:     parameters.NumOfAgents,
	AgentHP:        parameters.AgentHP,
	dayInfo:        parameters.DayInfo,
	healthInfo:     healthInfo,
	AgentsPerFloor: parameters.AgentsPerFloor,
	logger:         *stateLog.Logmanager.GetLogger("main").WithFields(log.Fields{"reporter": "simulation"}),
	stateLog:       stateLog,
	agentNewFuncs: map[agent.AgentType]AgentNewFunc{
		agent.Team1Agent1: agent1.New,
		agent.Team1Agent2: agent2.New,
	},
}
```
Which yielded
```
{"Current Simulation Tick":1,"level":"info","msg":"","reporter":"simulation","time":"2022-01-02T00:57:06Z"}
{"agent_id":"bea1e3a6-a8be-4a3f-aeb2-e16f89ad2dff","agent_type":"Team1Agent1","floor":1,"level":"info","message":"ProposeTreatyMessage","msg":"I sent a message","reporter":"agent","time":"2022-01-02T00:57:06Z"}
{"Current Simulation Tick":3,"level":"info","msg":"","reporter":"simulation","time":"2022-01-02T00:57:06Z"}
{"agent_id":"a7a84916-602f-41cf-a58a-13d74c09d1c2","agent_type":"Team1Agent2","level":"info","msg":"I have treaties","reporter":"agent","time":"2022-01-02T00:57:06Z"}
...
{"Current Simulation Tick":40,"level":"info","msg":"","reporter":"simulation","time":"2022-01-02T00:57:06Z"}
{"agent_id":"a7a84916-602f-41cf-a58a-13d74c09d1c2","agent_type":"Team1Agent2","level":"info","msg":"I have treaties","reporter":"agent","time":"2022-01-02T00:57:06Z"}
{"level":"info","msg":"-----------------END----OF----DAY-----------------","reporter":"tower","time":"2022-01-02T00:57:06Z"}
{"agent_id":"a7a84916-602f-41cf-a58a-13d74c09d1c2","agent_type":"Team1Agent2","level":"info","msg":"I no longer have treaties","reporter":"agent","time":"2022-01-02T00:57:06Z"}
```

With reject as default
```
{"Current Simulation Tick":8,"level":"info","msg":"","reporter":"simulation","time":"2022-01-02T01:41:34Z"}
{"agent_id":"593e3e8f-3d91-4caf-a3f6-fefeaaacd25a","agent_type":"Team1Agent1","floor":3,"level":"info","message":"ProposeTreatyMessage","msg":"I sent a message","reporter":"agent","time":"2022-01-02T01:41:34Z"}
{"Current Simulation Tick":9,"level":"info","msg":"","reporter":"simulation","time":"2022-01-02T01:41:34Z"}
{"agent_id":"05c876c2-dee5-4476-a043-e1892358376f","agent_type":"Team1Agent2","level":"info","msg":"Rejected treaty","proposerFloor":3,"proposerID":"593e3e8f-3d91-4caf-a3f6-fefeaaacd25a","reporter":"agent","time":"2022-01-02T01:41:34Z","treatyID":"97d90d6c-c993-4d10-b763-c0af65d9520f"}
```

### CODE REVIEW

**Review by JaafarRammal — DISMISSED** [review]
Code / structure / arch LGTM, I'd prefer if you wait for Jason or Cyrus or Matt to review as well

**Comment by jzzheng22** [comment]
> Code / structure / arch LGTM, I'd prefer if you wait for Jason or Cyrus or Matt to review as well

I can't review it - I wrote some of the code

**Review by Silvertrousers — APPROVED** [review]
LGTM, has the required functionality, might want to be able to store the ID of the person you've agreed to the treaty with but I guess agent teams can handle that by looking at senderID

**Comment by JaafarRammal** [comment]
I think they can implement on their side things like custom count estimate and who the senders / receivers were
Thanks all!

File: pkg/infra/baseAgentMessages.go
Comment (JaafarRammal) [inline pkg/infra/baseAgentMessages.go:50]: I might have missed something but why does the Handle have something inside by default? It can just be empty?

Also I just realized, this structure means that `a.activeTreaties[msg.TreatyID()] = treaty` has to be written by the agent teams (just a note)

Comment (jzzheng22) [reply pkg/infra/baseAgentMessages.go:50]: It could just be empty yeah, but it's more that this should be the 'default' behaviour - agents can still override this.

Yeah... I don't like that everyone would have to write `a.activeTreaties[msg.TreatyID()] = treaty`  either... not sure how to fix. Happy to take suggestions.

Comment (woonmoon) [reply pkg/infra/baseAgentMessages.go:50]: Currently, it's the honest version of "reject every treaty that comes your way", so you let the sender know that you rejected. 
There's also the honest "accept every treaty that comes your way". 
These are the two extreme responses we probably want to test on honest runs. For more complicated functionality or for dishonest runs agent teams will override later I think.

**Review by Silvertrousers — APPROVED** [review]

**Pull request merged**
